### PR TITLE
feat: close the Drawer by dragging

### LINKED ISSUE #548: Close the `Drawer` by dragging
### Description

Currently, the `Drawer` closes by clicking outside or by clicking the `CloseButton`. We felt it would be good to place a draggable element on the edge of the `Drawer`, like the `iOS Drawer`, and have a process to close it when it is dragged for a certain amount of time.

### Problem Statement/Justification

On mobile, users want to be able to close the `Drawer` by dragging it from a usability perspective.

### Proposed Solution or API

Added `closeOnDrag` props.

```tsx
<Drawer isOpen={isOpen} onClose={onClose} closeOnDrag>
  {/* contents */}
</Drawer>
```

### Alternatives

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by yasunogithub**
Closes #548 

## Description

* Added closeOnDrag props.

### Known Issues
* Unable to select text when turning on the closeOnDrag option
* ~~After pulling the element in the correct direction, adding momentum in the opposite direction and releasing it may cause the force of inertia to break through the restriction.~~ fixed
https://github.com/yamada-ui/yamada-ui/assets/29519029/975c9cab-a9fa-4036-af8c-19ecfd9c403a

### CODE REVIEW

**Review by taroj1205 — CHANGES_REQUESTED** [review]
Hi yasunogithub,
Thank you for your work on this pull request. 
Could you please look into the second issue you have mentioned and consider implementing a fix?
> After pulling the element in the correct direction, adding momentum in the opposite direction and releasing it may cause the force of inertia to break through the restriction.

**Comment by yasunogithub** [comment]
> Hi yasunogithub, Thank you for your work on this pull request. Could you please look into the second issue you have mentioned and consider implementing a fix?
> 
> > After pulling the element in the correct direction, adding momentum in the opposite direction and releasing it may cause the force of inertia to break through the restriction.

I see. I'll try it.

**Comment by yasunogithub** [comment]
Hi @taroj1205 
I just have fixed. please check it. thanks.

**Review by taroj1205 — APPROVED** [review]
Thanks! Approved.

**Review by taroj1205 — CHANGES_REQUESTED** [review]
Hi @yasunogithub, 
Thank you for your PR.
Would you please be able to write tests for adding a new prop?

File: packages/components/modal/src/drawer.tsx [hunk 1]
```diff
@@ -49,7 +50,9 @@ export type DrawerProps = Omit<
   "scrollBehavior" | "animation" | "outside" | keyof ThemeProps
 > &
   ThemeProps<"Drawer"> &
-  DrawerOptions
+  DrawerOptions & {
+    closeOnDrag?: boolean
+  }
```
Comment (taroj1205) [inline packages/components/modal/src/drawer.tsx:55]:
```suggestion
  "scrollBehavior" | "animation" | "outside" | keyof ThemeProps
> &
  ThemeProps<"Drawer"> &
  DrawerOptions
```

`closeOnDrag?: boolean` goes to type DrawerOptions. Please also add JSDoc and default value.

File: packages/components/modal/src/drawer.tsx [hunk 2]
```diff
@@ -167,9 +173,58 @@ export const DrawerContent = forwardRef<DrawerContentProps, "div", false>(
...
+        case "top":
+          return info.velocity.y <= -100 || info.offset.y <= -80
+        case "bottom":
+          return info.velocity.y >= 100 || info.offset.y >= 80
+        case "left":
+          return info.velocity.x <= -100 || info.offset.x <= -80
+        case "right":
+          return info.velocity.x >= 100 || info.offset.x >= 80
+      }
+    }
```
Comment (taroj1205) [inline packages/components/modal/src/drawer.tsx:213]: Please use `useCallback`.

File: .changeset/early-mirrors-fry.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+"@yamada-ui/modal": minor
+"@yamada-ui/use-pan-event": patch
+---
+
+Close the Drawer by dragging
```
Comment (taroj1205) [inline .changeset/early-mirrors-fry.md:6]: Please separate the changeset (modal and use-pan-event).

File: packages/components/modal/src/drawer.tsx [hunk 4]
```diff
@@ -99,7 +102,6 @@ export const Drawer = forwardRef<DrawerProps, "div">(
       validChildren,
       DrawerOverlay,
     )
-
```
Comment (taroj1205) [inline packages/components/modal/src/drawer.tsx:102]: Please do not delete this line.

File: packages/components/modal/src/drawer.tsx (hunk 2 above)
Comment (yasunogithub) [reply packages/components/modal/src/drawer.tsx:213]: I see. I'll try it.

Comment (yasunogithub) [reply packages/components/modal/src/drawer.tsx:213]: Is it right?
```javascript
    const getDragDirectionRestriction = useCallback(() => {
      switch (placement) {
        case "top":
          return { bottom: 0 }
        case "bottom":
          return { top: 0 }
        case "left":
          return { right: 0 }
        case "right":
          return { left: 0 }
      }
    }, [placement]);

    const getDragDirection = useCallback(() => {
      switch (placement) {
        case "top":
        case "bottom":
          return "y"
        case "left":
        case "right":
          return "x"
      }
    }, [placement]);

    const isCloseByDragInfo = useCallback((info: PanEventInfo) => {
      switch (placement) {
        case "top":
          return info.velocity.y <= -100 || info.offset.y <= -80
        case "bottom":
          return info.velocity.y >= 100 || info.offset.y >= 80
        case "left":
          return info.velocity.x <= -100 || info.offset.x <= -80
        case "right":
          return info.velocity.x >= 100 || info.offset.x >= 80
      }
    }, [placement]);
```

File: .changeset/early-mirrors-fry.md (hunk 3 above)
Comment (yasunogithub) [reply .changeset/early-mirrors-fry.md:6]: @taroj1205 
sorry, how can I separate changeset?

**Comment by yasunogithub** [comment]
> Hi @yasunogithub, Thank you for your PR. Would you please be able to write tests for adding a new prop?

@taroj1205 
Do I need to create tests/drawer.test.tsx? What items should I test?

**Comment by taroj1205** [comment]
> > Hi @yasunogithub, Thank you for your PR. Would you please be able to write tests for adding a new prop?
> 
> @taroj1205 Do I need to create tests/drawer.test.tsx? What items should I test?

Yes, please create `tests/drawer.test.tsx`. You can read https://github.com/yamada-ui/yamada-ui/pull/937/files#diff-7501cf4213f8d297d07adea960b822cfdc43e13fcc9a3130b338b8b07107f22e to help you.
You should test the lines you have added/changed with this PR including the new functions you have added.
To see if the tests are working, please run `pnpm test:coverage`, go to the files you have changed and see if the lines are no longer highlighted in red.

File: .changeset/early-mirrors-fry.md (hunk 3 above)
Comment (taroj1205) [reply .changeset/early-mirrors-fry.md:6]: You can remove one of the either like this.
```diff
---
"@yamada-ui/modal": minor
--"@yamada-ui/use-pan-event": patch
---
```

After you have removed the line, please run `pnpm changeset` and generate changeset for `@yamada-ui/use-pan-event`

File: packages/components/modal/src/drawer.tsx (hunk 2 above)
Comment (taroj1205) [reply packages/components/modal/src/drawer.tsx:213]: Yes, if closeOnDrag still work properly with this, this would be correct.

**Review by setodeve — APPROVED** [review]

**Review by taroj1205 — APPROVED** [review]

**Pull request merged**